src/ImageSharp.Drawing/Processing/Drawing/Brushes/LinearGradientBrush.cs

```diff
@@ -0,0 +1,211 @@
+﻿using System;
+using System.Numerics;
+
+using SixLabors.ImageSharp.Advanced;
+using SixLabors.ImageSharp.Memory;
+using SixLabors.ImageSharp.PixelFormats;
+using SixLabors.ImageSharp.PixelFormats.PixelBlenders;
+using SixLabors.Primitives;
+
+namespace SixLabors.ImageSharp.Processing.Drawing.Brushes
+{
+    /// <summary>
+    /// Provides an implementation of a brush for painting gradients within areas.
+    /// Supported right now:
+    /// - a set of colors in relative distances to each other.
+    /// - two points to gradient along.
+    /// </summary>
+    /// <typeparam name="TPixel">The pixel format</typeparam>
+    public class LinearGradientBrush<TPixel> : IBrush<TPixel>
+        where TPixel : struct, IPixel<TPixel>
+    {
+        private readonly Point p1;
+
+        private readonly Point p2;
+
+        private readonly Tuple<float, TPixel>[] keyColors;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LinearGradientBrush{TPixel}"/> class.
+        /// </summary>
+        /// <param name="p1">Start point</param>
+        /// <param name="p2">End point</param>
+        /// <param name="keyColors">a set of color keys and where they are. The double must be in range [0..1] and is relative between p1 and p2.</param>
+        public LinearGradientBrush(Point p1, Point p2, params Tuple<float, TPixel>[] keyColors)
+        {
+            this.p1 = p1;
+            this.p2 = p2;
+            this.keyColors = keyColors;
+        }
+
+        /// <inheritdoc />
+        public BrushApplicator<TPixel> CreateApplicator(ImageFrame<TPixel> source, RectangleF region, GraphicsOptions options)
+            => new LinearGradientBrushApplicator(source, this.p1, this.p2, this.keyColors, region, options);
+
+        /// <summary>
+        /// The linear gradient brush applicator.
+        /// </summary>
+        private class LinearGradientBrushApplicator : BrushApplicator<TPixel>
+        {
+            private readonly Point start;
+
+            private readonly Point end;
+
+            private readonly Tuple<float, TPixel>[] colorStops;
```

Comment:
I suggest to define a small private struct instead of `Tuple<float, TPixel>`. It provides:
- Better readability
- Better memory locality (`Tuple<>`-s are classes on heap)

Reply:
good point. Thanks. Except that the struct can't be private as it's used as a parameter in the Brush constructor. Therefore I'll use a public nested struct instead.